### PR TITLE
Add support for IAsyncEnumerable<T> where T is value type

### LINKED ISSUE #17139: Web API engine doesn't handle IAsyncEnumerable<int> type as expected.
I encountered an interesting behavior while exploring IAsyncEnumerable in an ASP.NET Web API project. Consider the following code samples:
Sample 1 (int array) returns {} as JSON result.

### To Reproduce
```

    // Code Sample 1
    [HttpGet]
    public async IAsyncEnumerable<int> GetAsync()
    {
        for (int i = 0; i < 10; i++)
        {
            await Task.Delay(1000);
            yield return i;
        }
    }


    // Code Sample 2
    [HttpGet]
    public async IAsyncEnumerable<string> GetAsync()
    {
        for (int i = 0; i < 10; i++)
        {
            await Task.Delay(1000);
            yield return i.ToString();
        }
    }


```
### Further technical details
- ASP.NET Core version : 3.0.0
- Include the output of `dotnet --info`
.NET Core SDK (reflecting any global.json):
 Version:   3.0.100
 Commit:    04339c3a26

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17763
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.0.100\

Host (useful for support):
  Version: 3.0.0
  Commit:  7d57652f33

- The IDE : VS 2019 Community

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes https://github.com/aspnet/AspNetCore/issues/17139

### CODE REVIEW

File: src/Mvc/Mvc.Core/src/Infrastructure/AsyncEnumerableReader.cs [hunk 1]
```diff
@@ -48,37 +45,39 @@ public AsyncEnumerableReader(MvcOptions mvcOptions)
...
+                    _asyncEnumerableConverters.TryAdd(type, reader);
+                }
             }
 
-            return result(value);
+            return reader != null;
         }
 
-        private async Task<ICollection> ReadInternal<T>(IAsyncEnumerable<object> value)
+        private async Task<ICollection> ReadInternal<T>(object value)
```
Comment (rynowak) [inline src/Mvc/Mvc.Core/src/Infrastructure/AsyncEnumerableReader.cs]: This is way smaller of a change than I expected.

File: src/Mvc/Mvc.Core/test/Infrastructure/AsyncEnumerableReaderTest.cs [hunk 2]
```diff
@@ -5,57 +5,168 @@
...
-            var collection = Assert.IsAssignableFrom<ICollection<string>>(result);
+            Assert.True(result);
+            var readCollection = await reader(asyncEnumerable);
+            var collection = Assert.IsAssignableFrom<ICollection<string>>(readCollection);
             Assert.Equal(new[] { "0", "1", "2", }, collection);
         }
 
         [Fact]
-        public async Task ReadAsync_ReadsIAsyncEnumerable_ImplementingMultipleAsyncEnumerableInterfaces()
+        public async Task TryGetReader_ReturnsReaderForIAsyncEnumerableOfValueType()
```
Comment (rynowak) [inline src/Mvc/Mvc.Core/test/Infrastructure/AsyncEnumerableReaderTest.cs]: 🎆 👏 👏

File: src/Mvc/Mvc.Core/test/Infrastructure/AsyncEnumerableReaderTest.cs [hunk 3]
```diff
@@ -5,57 +5,168 @@
...
+
+            // Assert
+            Assert.True(result);
+            var readCollection = await reader(asyncEnumerable);
+            var collection = Assert.IsAssignableFrom<ICollection<int>>(readCollection);
+            Assert.Equal(new[] { 0, 1, 2, }, collection);
+        }
+
+        [Fact]
+        public void TryGetReader_ReturnsCachedDelegate()
```
Comment (rynowak) [inline src/Mvc/Mvc.Core/test/Infrastructure/AsyncEnumerableReaderTest.cs]: How does this behave when a type implements multiple IAsyncEnumerables<>

**Review by rynowak — APPROVED** [review]

File: src/Mvc/Mvc.Core/test/Infrastructure/AsyncEnumerableReaderTest.cs (hunk 3 above)
Comment (pranavkm) [reply src/Mvc/Mvc.Core/test/Infrastructure/AsyncEnumerableReaderTest.cs]: We cache this based on the passed in type, so it just works. I added a test for good measure.

**Pull request merged**